src/common/tls.c

```diff
@@ -0,0 +1,227 @@
+/**
+ * (C) Copyright 2016-2023 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * It implements thread-local storage (TLS) for DAOS.
+ */
+#include <pthread.h>
+#include <daos/tls.h>
+
+/* The array remember all of registered module keys on one node. */
+static struct daos_module_key *daos_module_keys[DAOS_MODULE_KEYS_NR] = {NULL};
+pthread_mutex_t                daos_module_keys_lock                 = PTHREAD_MUTEX_INITIALIZER;
+
+static __thread bool           dc_tls_thread_init;
+
+static pthread_key_t           dss_tls_key;
+static pthread_key_t           dc_tls_key;
+
+void
+daos_register_key(struct daos_module_key *key)
+{
+	int i;
+
+	D_MUTEX_LOCK(&daos_module_keys_lock);
```

Comment:
In all honesty, it would probably be best for you to just go ahead and submit patches for these kinds of improvements. I have no strong ownership of the engine/client side of this beyond what I needed to do to make it work with the agent, and I certainly wouldn't claim to know that the current code/approach is the best possible implementation.